lib/cchain/provider/cache.go

```diff
@@ -0,0 +1,86 @@
+package provider
+
+import (
+	"sync"
+
+	"github.com/omni-network/omni/lib/xchain"
+)
+
+// chainVersionCache contains approved attestations of a chain version with all heights above the watermark.
+type chainVersionCache struct {
+	cache     map[uint64]xchain.Attestation
+	watermark uint64
+}
+
+// attestationCache if a thread-safe collection of caches containing attestations per chain version.
+// All cache sizes will be kept under `cacheSize` elements.
+type attestationCache struct {
+	cacheSizes int
+	caches     map[xchain.ChainVersion]*chainVersionCache
```

Comment:
I'd just using `sync.Map` here and then call `sync.LoadOrStore` below, this avoids having a single lock on all caches, then the lock can moved inside each cache, slightly more performant

Reply:
Oh great, not sure I knew about `sync.Map` 👍

Reply:
I remember now why I need the global lock. The `caches` itself is mutable and shared too. Unless we can pre-initialize upfront for all known chain versions.